Hour charting

cryogenian opened the pull request:
Fixes #1038 

+ AFAIK, we're going to remove `axes` field from `TaggedResource`  --> we don't check axes for chart type --> first problem solved automatically
+ Changed data extraction, and added date and datetime semantics. So, at least _CASE_ with `"01:00"` works now. 

@garyb please take a look

garyb commented on src/SlamData/Workspace/Card/BuildChart/Line/Eval.purs:
```diff
@@ -210,16 +211,16 @@ buildLine r records axes = do
+    Ax.Time → {axisType: ET.Category, interval: Just 0, heightMult: 2}
+    Ax.Date → {axisType: ET.Time, interval: Just 0, heightMult: 2}
+    Ax.DateTime → {axisType: ET.Time, interval: Just 0, heightMult: 2}
+    Ax.Category → {axisType: ET.Category, interval: Just 0, heightMult: 1}
```
This function seems to be duplicated in a bunch of places, unless it subtly differs for the various chart types can we move it out into a common module somewhere?

cryogenian replied:
Will do

garyb commented on src/SlamData/Workspace/Card/BuildChart/Radar/Eval.purs:
```diff
@@ -75,12 +75,14 @@ buildRadarData r records = series
     → Maybe String >> Maybe String >> String >> Array Number
   dataMapFoldFn acc js =
-    case toString =<< cursorGet r.category js of
+    case map printSemantics $ analyzeJson =<< cursorGet r.category js of
```
Likewise, can we make a general `getKey` or something, since there's a lot of repetition of this chain of expressions. I guess the general form would be more like the other case:

```
getKey js cursors = map printSemantics $ analyzeJson =<< flip cursorGet js =<< cursors
```

So you could just `getKey js (pure r.category)` or something here, and use it as-is for the others.

garyb commented on src/SlamData/Workspace/Card/BuildChart/Semantics.purs:
```diff
@@ -38,116 +39,57 @@ data Semantics
+    month = fromEnum $ Dd.month date
+    day = fromEnum $ Dd.day date
+  in
+    show year ⊕ "-" ⊕ show month ⊕ "-" ⊕ show day
```
Can we use `purescript-formatters` here instead now?

cryogenian replied:
Yeah, I'm a bit unsure about performance though. This function is called _many_ times.

garyb commented on src/SlamData/Workspace/Card/BuildChart/Semantics.purs:
```diff
@@ -196,13 +138,13 @@ instance ordSemantics ∷ Ord Semantics where
   compare (Bool b) (Bool b') = compare b b'
   compare (Bool _) _ = LT
```
I'm pretty sure `Ord` and `Eq` should be derivable for `Semantics`.

garyb commented on src/SlamData/Workspace/Card/BuildChart/Semantics.purs:
```diff
@@ -335,19 +304,35 @@ instance encodeJsonSemantics ∷ EncodeJson Semantics where
+  encodeJson s@(Time t)
      = "type" := "time"
-     ~> "value" := t
+     ~> "value" := printSemantics s
```
Is this change to the `Time` case going to cause issues with backwards compatible encode/decode?

cryogenian replied:
Nope, actually this codecs aren't used anywhere :)

cryogenian commented:
+ Extracted `getMaybeString` and `getValues` 
+ Using `derive instance` for `Semantics`
+ Removed `Encode/DecodeJson` instances for `Semantics` 
+ Added script replacing `CRLF -> LF` and call it in each script in `package.json` (I can remove it if its too noisy) 
+ Added comment for `printSemantics`

garyb commented on package.json:
```diff
@@ -5,7 +5,7 @@
     "url": "https://github.com/slamdata/slamdata"
   },
-  "version": "v3.1.0-rc.2",
+  "version": "v3.0.11",
```
Oops? :smile:

cryogenian replied:
Oops :D

cryogenian commented on package.json:
```diff
@@ -15,11 +15,11 @@
     "property-tests": "gulp bundle-property-tests && node tmp/js/property-tests",
     "test": "gulp bundle-test && node test",
-    "psa": "pulp build -I test/src --censor-lib --strict --stash"
+    "psa": "gulp replace-crlf && pulp build -I test/src --censor-lib --strict --stash"
```
I use cycle 
`pulp build -I test/src --censor-lib --stash` for start development
then 
`npm run psa` 
then 
`npm run psa && gulp bundle` 
then 
`npm run psa && npm run property-tests && npm run test` 
in every day work. (Yeah, pscid doesn't work on my machine) 

That's why I've added this thing here

garyb replied:
I tend to `npm run psa && gulp bundle` while working, and then run the "full" build with `add-headers` etc. before pushing a branch / PR.

You should be able to set your editor to default to LF too :smile: - I thought we had a `.editorconfig` file at one point too, that should dictate it for this project, but maybe I imagined it since it's not there now anyway.

cryogenian replied:
I have no idea why emacs sometimes shows CRLF as `^M` and sometimes doesn't. And why my git config doesn't change crlf -> lf on commit   :(

garyb replied:
It's probably down to whatever the `core.autocrlf` is set to: https://git-scm.com/book/en/v2/Customizing-Git-Git-Configuration#Formatting-and-Whitespace

cryogenian commented:
+ Duplicated headers
+ Version
+ Tests (for #1096 also)
